### PR TITLE
Implement materials cherry-picking duplication

### PR DESCRIPTION
**Pull request by kxmbrian**
v1 only allows duplication of top-level folder. This implementation allows users to select which folders and materials they would like to duplicate.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -44,7 +44,7 @@ node_modules
 
 # Ignore generated js bundles
 /app/assets/webpack/*
```
Comment (jeremyyap) [inline .gitignore]: I think you can remove this too, this was probably from react-on-rails

**Comment by allenwq** [comment]
@kxmbrian what if there're naming conflicts? i.e. folder or material with the same name exists in the target course.

**Comment by kxmbrian** [comment]
Shall we just warn the user on the 'checkout' screen, but let the user go ahead and do it?

**Comment by allenwq** [comment]
I was thinking maybe we should merge the folders with same name. e.g. 

Suppose we have 2 courses below:

Course A (current course):
Existing Materials: `Exams/1010FC/final.pdf`

Course B (target):
Existing Material: `Exams/1010S/final.pdf`

When the user select `final.pdf` and its folder(1010FC) in Course A.


In the above case, should the material and folder be copied to `Exams/1010FC/final.pdf` in course B?  What do you think?


> Shall we just warn the user on the 'checkout' screen, but let the user go ahead and do it?

I am fine with this too, can can support that later.

**Comment by kxmbrian** [comment]
I prefer to just warn the user and have the user manually move the items.

If I have:
```
Course 1010FC (current course):
Existing Materials: Exams/final.pdf, Exams/midterm.pdf, ...

Course 1010S (target):
Existing Materials: Exams/final.pdf, Exams/midterm.pdf, ...
```

and I want to eventually achieve:
```
Course 1010S (target):
Exams/final.pdf, Exams/midterm.pdf, ..., Exams(1010FC)/final.pdf, Exams(1010FC)/midterm.pdf, ...,
```

the user may have a hard time making corrections if s/he is unaware of the merging behavior.

**Comment by kxmbrian** [comment]
Pushed a commit to warn the user of naming conflicts

**Comment by weiqingtoh** [comment]
My 2 cents: 

I would prefer that we not use the course title (it is quite long for the programming ones). One of the more straightforward ways to handle naming conflicts is to just append a `(1)` behind (or `(2)` or `(3)`...). That is the default behavior I get on Chrome when downloading files of the same title. 

When we have time, we can extend from this behaviour by allowing other options.

**Comment by allenwq** [comment]
@kxmbrian in your example, the default behaviour that I proposed is:

```
# Finally after duplication:
Course 1010S (target):
Exams/final.pdf, Exams/midterm.pdf, ..., Exams/final(1).pdf, Exams/midterm(1).pdf, ...
```

We actually have a `#next_valid_name` method to achieve this in `Course::Material::Folder`, I think that can be extended to materials too.

> Pushed a commit to warn the user of naming conflicts

Will the user be able to run the duplication if there're naming conflicts?

**Comment by kxmbrian** [comment]
Added a commit that does the name de-conflicting for both duplicated folders and materials. Current behavior:

```
# Finally after duplication:
Course 1010S (target):
Exams/final.pdf, Exams/midterm.pdf, ..., Exams (0)/final.pdf, Exams (0)/midterm.pdf, ...
```

> I would prefer that we not use the course title (it is quite long for the programming ones)

In the example, the user wants to achieve this. The duplication module will not do it for him/her. Rather, the module will put it in the folder with the next unique name and let the user handle it from there. Would that work?

> Will the user be able to run the duplication if there're naming conflicts?

Yes

File: app/models/course/material/folder.rb [hunk 2]
```diff
@@ -106,6 +101,16 @@ def initialize_duplicate_children(duplicator, other)
                 map { |folder| duplicator.duplicate(folder) }
   end
 
+  def initialize_duplicate_materials(duplicator, other)
+    self.materials = if other.concrete?
```
Comment (weiqingtoh) [inline app/models/course/material/folder.rb]: Is it possible to explain why you introduced this - is it trying to protect the case where we accidentally try to duplicate non-concrete folders? Or are we allowing for duplication of non-concrete folders?

File: app/models/course/material.rb [hunk 3]
```diff
@@ -28,7 +30,11 @@ def record_userstamp
 
   def initialize_duplicate(duplicator, other)
     self.attachment = duplicator.duplicate(other.attachment)
-    self.folder = duplicator.duplicate(other.folder)
+    self.folder = if duplicator.duplicated?(other.folder)
```
Comment (weiqingtoh) [inline app/models/course/material.rb]: If I'm not wrong, you are relying on an implicit assumption that all concrete folders are to be duplicated before the materials are duplicated. I would rather that not be the case, is there a way to do this differently? Perhaps specifying the behavior more explicitly between `:course` and `:object` modes?

File: spec/services/course/duplication/object_duplication_service_spec.rb [hunk 4]
```diff
@@ -341,19 +341,19 @@
         end
         let!(:content) { create(:course_material, folder: folder) }
 
-        context 'when folders are duplicated but not their parents' do
+        context 'when folder is duplicated but not its parents nor materials' do
```
Comment (weiqingtoh) [inline spec/services/course/duplication/object_duplication_service_spec.rb]: For the test structure, I'm not sure if all the cases are covered. Would suggest to approach it in a more structured way, ie.
```
context 'when materials and folders are present' do
  context 'when no folders are selected for duplication' do
    it 'dups the selected materials into root folder' {}
  end

  context 'when folders are selected for duplication' do
    context 'when folders parents are selected for duplication' { it 'assigns the parent folder correctly' {} }
    context 'when folders parent are not selected for duplication' { it 'sets parent to root folder of target course' {} }
    context 'when folder materials are selected for duplication ' { it 'assigns the folders of materials correctly' {} }
  end
end
```

File: app/models/course/material/folder.rb (hunk 2 above)
Comment (allenwq) [reply app/models/course/material/folder.rb]: @kxmbrian  ```select { |material| duplicator.duplicated?(material) }``` does this means that materials will always be duplicated before folder?  what if the materials are duplicated later?

**Comment by kxmbrian** [comment]
The duplication will work whether a folder is duplicated before or after its associated material, as the tests assert. The same idea applies to all associated objects, e.g. categories, tabs, and assessments. Have added some documentation to the `duplicator` to clarify. A bit wordy - am open to suggestions on how to improve it.

Also re-nested the spec contexts:

```
when materials and folders are present
  when there are items with conflicting filenames
    gives the duplicated items unique names
  when a folder is duplicated
    when it does not have a containing folder
      duplicates it to destination course root folder
    when its containing folder is not duplicated
      duplicates it to destination course root folder
    when its containing folder is also duplicated
      when folder is duplicated after its parent
        associates them
      when folder is duplicated before its parent
        associates them
  when a material is duplicated
    when its containing folder is not duplicated
      duplicates it to target course root folder
    when its containing folder is also duplicated
      when material is duplicated before its folder
        associates them
      when material is duplicated after its folder
        associates them
```

Hope it's clear enough now!

File: app/models/course/material/folder.rb (hunk 2 above)
Comment (allenwq) [reply app/models/course/material/folder.rb]: @kxmbrian Can explain a bit more? since the order doesn't matter, why we need to check `duplicator.duplicated?(material)`?

**Comment by weiqingtoh** [comment]
I'm not sure why the [old discussion was hidden](https://github.com/Coursemology/coursemology2/pull/2583#discussion_r146172009), but want to raise it again: 

```ruby
def initialize_duplicate(duplicator, other)
    self.attachment = duplicator.duplicate(other.attachment)
    self.folder = if duplicator.duplicated?(other.folder)
                    duplicator.duplicate(other.folder)
                  else
                    duplicator.options[:target_course].root_folder
                  end
    self.updated_at = other.updated_at
    self.created_at = other.created_at
    set_duplication_flag
  end
```

For the `self.folder` assignment, my general suggestion is that although the code works, it relies on implicit assumptions (ie. course duplication mode, folder will be duplicated before material). In this case, I rather be slightly more verbose to provide readable code. Hence, something along these lines: 

```ruby
self.folder = initialise_duplicate_folder

...

def initialise_duplicate_folder
  has_parent_folder = duplicator.mode == :course || (duplicator.mode == :object && duplicator.duplicated?(other.folder)
  if has_parent_folder
    duplicator.duplicate(other.folder)
  else
    duplicator.options[:target_course].root_folder
  end
end
```

This will probably be useful if for some reason we need to refactor parts of Duplication. Thoughts?

**Comment by kxmbrian** [comment]
Sorry for the confusion. The code does *not* rely on the assumption that folder will be duplicated before material. Not sure if the explanation in the added commit `Improve duplicator documentation` was sufficient, but I've added more comments in the `material.rb` and `folder.rb`.

The key thing to note is that if a parent folder has not been duplicated yet but will be duplicated later, then the child item is parked under the root folder only *temporarily*.

**Comment by jeremyyap** [comment]
Reference #2007 and #2168
